### ISSUE TITLE
`to_if_alone` events firing in the wrong order, some events being dropped

### ISSUE THREAD

**Issue by wincent** [tekezo/Karabiner-Elements#789]
So after a few days of testing v0.91.4 I've found that the issue reported in #785 — or more accurately, a variant of it — is still present, it is just happening much less frequently.

Here is an event log of me pressing the "Enter" key about 7 times. I've put a blank line in between each press to make it clearer:

```
eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:

eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:

eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:

eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:

eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:

eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:

eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:

eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:
```

Two interesting things to note here:

- Events are being logged as `KeyUp` then `KeyDown`; not sure if this is a display issue only or they really are firing out of order. This happens for any key with a `to_if_alone` config. Compare that with a normal keypress, which logs as `KeyDown` then `KeyUp` as expected:

```
eventType:KeyDown         code:0x0        name:a               flags:                                   misc:
eventType:KeyUp           code:0x0        name:a               flags:                                   misc:
```

- The other thing to note in the log for `return_or_enter` is that the final event *only* shows a `KeyUp` and no `KeyDown`. When I'm typing I feel this as an unresponsive key (no "enter" actually gets produced, so I have to press it again).

I can reproduce this issue on the other side of my keyboard with a `to_if_alone` config where I have Caps Lock set to Backspace alone and Control otherwise:

```
eventType:FlagsChanged    code:0x3b       name:left_control    flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3b       name:left_control    flags:                                   misc:
eventType:KeyUp           code:0x33       name:delete_or_backs flags:                                   misc:
eventType:KeyDown         code:0x33       name:delete_or_backs flags:                                   misc:

eventType:FlagsChanged    code:0x3b       name:left_control    flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3b       name:left_control    flags:                                   misc:
eventType:KeyUp           code:0x33       name:delete_or_backs flags:                                   misc:
eventType:KeyDown         code:0x33       name:delete_or_backs flags:                                   misc:

eventType:FlagsChanged    code:0x3b       name:left_control    flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3b       name:left_control    flags:                                   misc:
eventType:KeyUp           code:0x33       name:delete_or_backs flags:                                   misc:
```

I'd say this is happening on average somewhere between 1 in 10 and 1 in 20 keypresses.

**Comment by wincent** [comment]
I notice that when a `KeyDown` event gets dropped, on the next keypress I see the `KeyDown` alone, but no `KeyUp`. This sample corresponds to 4 key presses, with press 1, 3, and 4 all producing an effect visible to applications, but 2 getting swallowed:

```
// (1) Get Up/Down pair.
eventType:FlagsChanged    code:0x3b       name:left_control    flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3b       name:left_control    flags:                                   misc:
eventType:KeyUp           code:0x33       name:delete_or_backs flags:                                   misc:
eventType:KeyDown         code:0x33       name:delete_or_backs flags:                                   misc:

// (2) Only get Up.
eventType:FlagsChanged    code:0x3b       name:left_control    flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3b       name:left_control    flags:                                   misc:
eventType:KeyUp           code:0x33       name:delete_or_backs flags:                                   misc:

// (3) Only get Down.
eventType:FlagsChanged    code:0x3b       name:left_control    flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3b       name:left_control    flags:                                   misc:
eventType:KeyDown         code:0x33       name:delete_or_backs flags:                                   misc:

// (4) Get Up/Down pair.
eventType:FlagsChanged    code:0x3b       name:left_control    flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3b       name:left_control    flags:                                   misc:
eventType:KeyUp           code:0x33       name:delete_or_backs flags:                                   misc:
eventType:KeyDown         code:0x33       name:delete_or_backs flags:                                   misc:
```

Same on the "Caps Lock" side:

```
eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:

eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:

eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:

eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:
```

**Comment by wincent** [comment]
Here is my attempt at constructing a failing test to demonstrate the order issue, but it doesn't fail. (Sorry, not familiar with the code base, so still not sure if this is a display-only issue in the event viewer or the events really are coming through in reverse):

```diff
diff --git a/tests/src/post_event_to_virtual_devices/json/alone.json b/tests/src/post_event_to_virtual_devices/json/alone.json
index b9779fc..217ed9c 100644
--- a/tests/src/post_event_to_virtual_devices/json/alone.json
+++ b/tests/src/post_event_to_virtual_devices/json/alone.json
@@ -89,5 +89,27 @@
             }
         ],
         "type": "basic"
+    },
+    {
+        "description": "return_or_enter",
+        "from": {
+            "key_code": "return_or_enter",
+            "modifiers": {
+                "optional": [
+                    "any"
+                ]
+            }
+        },
+        "to": [
+            {
+                "key_code": "right_control"
+            }
+        ],
+        "to_if_alone": [
+            {
+                "key_code": "return_or_enter"
+            }
+        ],
+        "type": "basic"
     }
 ]
diff --git a/tests/src/post_event_to_virtual_devices/test.cpp b/tests/src/post_event_to_virtual_devices/test.cpp
index 103a449..e1a70d0 100644
--- a/tests/src/post_event_to_virtual_devices/test.cpp
+++ b/tests/src/post_event_to_virtual_devices/test.cpp
@@ -1164,6 +1164,27 @@ TEST_CASE("actual examples") {
 
     time_stamp = 0;
 
+    ENQUEUE_EVENT(helper.get_input_event_queue(), 1, time_stamp += interval, return_or_enter_event, key_down);
+    ENQUEUE_EVENT(helper.get_input_event_queue(), 1, time_stamp += interval, return_or_enter_event, key_up);
+
+    helper.manipulate();
+
+    time_stamp = 0;
+    expected.clear();
+
+    ENQUEUE_KEYBOARD_EVENT(expected, right_control, 1, time_stamp);
+    ENQUEUE_KEYBOARD_EVENT(expected, right_control, 0, time_stamp);
+    ENQUEUE_KEYBOARD_EVENT(expected, return_or_enter, 1, time_stamp);
+    ENQUEUE_KEYBOARD_EVENT(expected, return_or_enter, 0, time_stamp);
+
+    REQUIRE(helper.get_events() == expected);
+  }
+
+  {
+    actual_examples_helper helper("alone.json");
+
+    time_stamp = 0;
+
     ENQUEUE_EVENT(helper.get_input_event_queue(), 1, time_stamp += interval, right_control_event, key_down);
     ENQUEUE_EVENT(helper.get_input_event_queue(), 1, time_stamp += interval, f1_event, key_down);
     ENQUEUE_EVENT(helper.get_input_event_queue(), 1, time_stamp += interval, right_control_event, key_up);
```

**Comment by wincent** [comment]
I added some `logger` calls to the grabber to see what events it was grabbing and what events it was dispatching to the virtual device when this bug occurs. The output looks like this both when the bug happens and when it doesn't:

```
[2017-06-22 00:00:50.554] [info] [grabber] down (return down)
[2017-06-22 00:00:50.554] [info] [grabber] dispatch 1 (control down)
[2017-06-22 00:00:50.641] [info] [grabber] up (return up)
[2017-06-22 00:00:50.641] [info] [grabber] sent key down (alone) at 3868145416768
[2017-06-22 00:00:50.641] [info] [grabber] sent key up (alone) at 3868145416769
[2017-06-22 00:00:50.641] [info] [grabber] down (synthetic return down)
[2017-06-22 00:00:50.641] [info] [grabber] up (synthetic return up)
[2017-06-22 00:00:50.641] [info] [grabber] dispatch 0 (control up)
[2017-06-22 00:00:50.643] [info] [grabber] dispatch 1 (return down)
[2017-06-22 00:00:50.643] [info] [grabber] dispatch 0 (return up)
```

Basically, this shows that the grabber *is* seeing all the incoming events, and it's also dispatching the corresponding events to the virtual device in the right order. So, somehow, by the time the events get to the event viewer, Apple has switched the order of the "up" and "down", and it is also occasionally entirely swallowing events.

**Comment by tekezo** [comment]
Thank you for feedback.
I think it's an instability in the window system layer.

We have to increasing the wait around modifier events to enough.
(But it's causes a delay, so I want to set it least value.)

https://github.com/tekezo/Karabiner-Elements/blob/master/src/core/grabber/include/manipulator/details/post_event_to_virtual_devices.hpp#L163

For example:
```
      // wait is 5 milliseconds
      auto wait = time_utility::nano_to_absolute(NSEC_PER_MSEC * 5);
```

**Comment by wincent** [comment]
**tl;dr:** any use of the Hammerspoon hs.eventtap module will trigger this problem; remove use of the module and the problem goes away.

-----

Actually, I'd already tried that (the 5ms delay), so I instead tried a 10ms delay with no success; seems no delay is sufficient. So I also tried inserting a bigger delay between the key down and key up (1ms, up from 1ns) but that also had no effect.

I noticed that this morning instead of seeing only "key up" followed by only "key down"  when the bug manifests, I was seeing only "key down", so clearly there was something even more non-deterministic going on. Here is an example with 4 presses:

```
// (1) Working Up/Down pair.
eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:

// (2) Working Down/Up pair (order is different!).
eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:

// (3) Broken (Down only).
eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:

// (4) Working Up/Down pair.
eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:
```

Finally, I noticed that consistently, when you start Karabiner-Elements, the first key presses for any "to_if_alone" keys look like these:

```
// (1) First press of return, works, but only shows "Down".
eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:

// (4) Second press of return, works, and shows "Up/Down".
eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:

// (3) First press of Caps Lock, works, but only shows "Down".
eventType:FlagsChanged    code:0x3b       name:left_control    flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3b       name:left_control    flags:                                   misc:
eventType:KeyDown         code:0x33       name:delete_or_backs flags:                                   misc:

// (4) Second press of Caps Lock, works, and shows "Up/Down".
eventType:FlagsChanged    code:0x3b       name:left_control    flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3b       name:left_control    flags:                                   misc:
eventType:KeyUp           code:0x33       name:delete_or_backs flags:                                   misc:
eventType:KeyDown         code:0x33       name:delete_or_backs flags:                                   misc:
```

Finally, on a hunch, I stopped Hammerspoon and saw that the event orders were suddenly correct always!

```
eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:

eventType:FlagsChanged    code:0x3e       name:right_control   flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3e       name:right_control   flags:                                   misc:
eventType:KeyDown         code:0x24       name:return_or_enter flags:                                   misc:
eventType:KeyUp           code:0x24       name:return_or_enter flags:                                   misc:

eventType:FlagsChanged    code:0x3b       name:left_control    flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3b       name:left_control    flags:                                   misc:
eventType:KeyDown         code:0x33       name:delete_or_backs flags:                                   misc:
eventType:KeyUp           code:0x33       name:delete_or_backs flags:                                   misc:

eventType:FlagsChanged    code:0x3b       name:left_control    flags:Ctrl                               misc:
eventType:FlagsChanged    code:0x3b       name:left_control    flags:                                   misc:
eventType:KeyDown         code:0x33       name:delete_or_backs flags:                                   misc:
eventType:KeyUp           code:0x33       name:delete_or_backs flags:                                   misc:
```

I have/had Hammerspoon running with what I thought was [a passive `hs.eventtap` module to gather aggregate keyboard statistics](https://github.com/wincent/wincent/blob/29624a7b37fbe9f87860f058ad63b5f42d16c2d7/roles/dotfiles/files/.hammerspoon/keylogger.lua). Even if I reduce it to a total no-op, this breaks the events for Karabiner-Elements:

```
local keyDown = hs.eventtap.event.types.keyDown

local keyHandler = (function(evt)
  -- The docs (http://www.hammerspoon.org/docs/hs.eventtap.html#new)
  -- say that returning false will allow the event to "propagate to any other
  -- applications watching for that event".
  return false
end)

hs.eventtap.new({keyDown}, keyHandler):start())
```

I can toggle the bug on and off by calling `start()` and `stop()` on the eventtap. So, this is probably either a Hammerspoon bug, or an Apple bug (where any use at all of the event tap API breaks Karabiner).

Going to close this as I have a workaround (don't use `hs.eventtap`). Sorry for the noise!

**Comment by asmagill** [comment]
As one of the contributors to Hammerspoon, I wanted to chime in -- it appears that your Hammerspoon event watcher is watching only for keyDown while your Karabiner-Elements debug log shows events keyDown, keyUp, and flagsChanged.  By setting up Hammerspoon to only capture one of these events, the path that the specific event takes through the system is now one step longer than the path the others take... I'd be curious to see if modifying the Hammerspoon snippet to capture all three events (even if you just return false immediately) returns K-E to receiving the events in the expected order.

**Comment by wincent** [comment]
Sounds like a credible theory for sure, @asmagill.

**Comment by cwagner22** [comment]
I have the same issue even without hammerspoon running. Same logs but weirdly I only notice the issue when I'm using Alfred or Kodi. In Kodi I have to press any other key after releasing enter to register it or else nothing happens. I'm using karabiner elements 0.91.13
Any help?
